feat: converted src/APIs to TS

shaikahmadnawaz opened the pull request:
## Describe your changes
This PR converts the `src/APIs` directory to TypeScript by updating the file extensions from `.js` to `.ts`. This change helps improve the maintainability and type safety of the codebase. No logical changes have been made to the files; only the file extensions and necessary type updates have been applied.

## Issue ticket number and link
Closed #673

## Motivation and Context
The motivation behind this change is to enhance the codebase by leveraging TypeScript's static typing and improved developer tooling. By converting the src/APIs files to TypeScript, we can catch potential type errors during development and improve the overall robustness of the code.

## How Has This Been Tested?
No specific tests have been performed for this PR since it primarily involves file extension changes. However, the codebase should be tested after merging this PR to ensure that the application continues to function correctly.

## Screenshots (if appropriate):
N/A (No visual changes have been made)

Please review and merge this PR.

shaikahmadnawaz committed:
feat: converted src/APIs to TS

ijemmao commented:
@shaikahmadnawaz great! can you confirm your changes are working by running `yarn test` and then attaching a screenshot of the passing test in your PR description?

shaikahmadnawaz commented:
> @shaikahmadnawaz great! can you confirm your changes are working by running `yarn test` and then attaching a screenshot of the passing test in your PR description?

I got this 👇, check once, how can I pass all tests?

![Screenshot 2023-06-21 181508](https://github.com/nkowaokwu/igbo_api/assets/96189881/ad204a8d-aff4-4aa8-9224-11582ee089bf)

ijemmao commented:
@shaikahmadnawaz there's no need for an env variable - what OS are you running on?

also to confirm, you have MongoDB locally on your machine and Java?

ijemmao committed:
Merge branch 'master' into api/js-ts

ijemmao commented:
@shaikahmadnawaz looks like the pipeline is failing due to a new TS change made in this PR - https://github.com/nkowaokwu/igbo_api/actions/runs/5340617462/jobs/9680629966?pr=677

shaikahmadnawaz committed:
feat: fixed type mismatch

shaikahmadnawaz committed:
Merge branch 'api/js-ts' of https://github.com/shaikahmadnawaz/igbo_api into api/js-ts

shaikahmadnawaz commented:
> @shaikahmadnawaz great! can you confirm your changes are working by running `yarn test` and then attaching a screenshot of the passing test in your PR description?

![Screenshot 2023-06-22 074627](https://github.com/nkowaokwu/igbo_api/assets/96189881/a8acbda4-e79d-402f-9d9a-39fed93e92c2)

shaikahmadnawaz commented:
One suggestion from my side, check whether this both APIs are working well before merging it to master branch.

ijemmao commented:
@shaikahmadnawaz the build step in GitHub is failing as expected since the type checking is failing - you'll need to make the appropriate change to allow TS to pass during the build step - the source of truth to ensure that this change is correct is our GitHub testing pipeline on this PR

shaikahmadnawaz commented:
> @shaikahmadnawaz the build step in GitHub is failing as expected since the type checking is failing - you'll need to make the appropriate change to allow TS to pass during the build step - the source of truth to ensure that this change is correct is our GitHub testing pipeline on this PR

Okay, I will work on it.

ukor commented on src/APIs/FlagsAPI.ts:
```diff
@@ -0,0 +1,56 @@
+}: HandleWordFlagsParams) => {
+  const updatedWords = compact(
+    words.map((word) => {
+      let updatedWord = assign({}, word);
```
Is there a reason why the Loadash is used instead of the native `Object.assign` method?

ukor commented on src/APIs/RedisAPI.ts:
```diff
@@ -11,49 +17,75 @@ export const getCachedWords = async ({ key, redisClient }) => {
-  const updatedData = assign(data);
+interface SetCachedWordsParams {
+  key: string;
+  data: any; // Update the type of `data` accordingly
```
`any` takes away the benefit of using typescript, avoid if possible

ukor replied:
Or better still use the `unknown` type

ijemmao commented:
@shaikahmadnawaz are you still working on this PR?

shaikahmadnawaz commented:
> @shaikahmadnawaz are you still working on this PR?

I don't know why it is not working.

ijemmao commented:
@shaikahmadnawaz when you see a failing CI test you can click on the details button to see the step where the CI failed:

![image](https://github.com/nkowaokwu/igbo_api/assets/16169291/2243188b-ede6-4411-a1c3-8cf09e852247)


**Here's the direct link to the error logs**: https://github.com/nkowaokwu/igbo_api/actions/runs/5340819225/jo

Looks like the RedisAPI is throwing the error saying that `redis` doesn't export `RedisClient`

## Error log
<img width="656" alt="Screen Shot 2023-07-03 at 10 46 51 PM" src="https://github.com/nkowaokwu/igbo_api/assets/16169291/8e6b9916-3c9b-4394-afa4-22f2ace9fd78">
bs/9690688217?pr=677

shaikahmadnawaz commented:
> @shaikahmadnawaz when you see a failing CI test you can click on the details button to see the step where the CI failed:
> 
> ![image](https://user-images.githubusercontent.com/16169291/250720982-2243188b-ede6-4411-a1c3-8cf09e852247.png)
> 
> **Here's the direct link to the error logs**: https://github.com/nkowaokwu/igbo_api/actions/runs/5340819225/jo
> 
> Looks like the RedisAPI is throwing the error saying that `redis` doesn't export `RedisClient`
> 
> ## Error log
> <img alt="Screen Shot 2023-07-03 at 10 46 51 PM" width="656" src="https://user-images.githubusercontent.com/16169291/250720776-8e6b9916-3c9b-4394-afa4-22f2ace9fd78.png">

Got it and I removed `RedisClient`

shaikahmadnawaz committed:
Update RedisAPI.ts

RedisAPI is throwing the error saying that redis doesn't export RedisClient, so I removed import of this.

zanio commented on src/APIs/RedisAPI.ts:
```diff
@@ -11,49 +16,75 @@ export const getCachedWords = async ({ key, redisClient }) => {
+  version: string; // Update the type of `version` accordingly
+}
+
+export const setCachedWords = async ({ key, data, redisClient, version }: SetCachedWordsParams) => {
+  const updatedData = assign({}, data);
   updatedData.words = minimizeWords(data.words, version);
   if (!redisClient.isFake) {
     await redisClient.set(key, JSON.stringify(updatedData), { EX: REDIS_CACHE_EXPIRATION });
   }
   return updatedData;
 };
```
I think explicitly stating the return types would give the full benefit of typescript.  The function returns and `updatedData` but there's no return type to it. The benefit of this is that where ever it called we would know what to expect

ijemmao replied:
hmm I'm not entirely convinced tbh, the type is inferred from `data` which should be defined within `SetCachedWordsParams`. explicitly defining the return type of the function is more for readability purposes, which you can argue is important in this case - but i can go either way

@shaikahmadnawaz can you do `assign(data)` instead?

zanio commented on src/APIs/RedisAPI.ts:
```diff
@@ -11,49 +16,75 @@ export const getCachedWords = async ({ key, redisClient }) => {
+  redisClient: RedisClient;
+}
+
+export const setCachedExamples = async ({ key, data, redisClient }: SetCachedExamplesParams) => {
   if (!redisClient.isFake) {
     await redisClient.set(key, JSON.stringify(data), { EX: REDIS_CACHE_EXPIRATION });
   }
   return data;
 };
```
Same for this

zanio commented on src/APIs/RedisAPI.ts:
```diff
@@ -11,49 +16,75 @@ export const getCachedWords = async ({ key, redisClient }) => {
+export const getAllCachedVerbsAndSuffixes = async ({ key, redisClient }: GetAllCachedVerbsAndSuffixesParams) => {
   const redisAllVerbsAndSuffixesKey = `verbs-and-suffixes-${key}`;
   const rawCachedAllVerbsAndSuffixes = await redisClient.get(redisAllVerbsAndSuffixesKey);
-  const cachedAllVerbsAndSuffixes = typeof rawCachedAllVerbsAndSuffixes === 'string'
-    ? JSON.parse(rawCachedAllVerbsAndSuffixes)
-    : rawCachedAllVerbsAndSuffixes;
+  const cachedAllVerbsAndSuffixes =
+    typeof rawCachedAllVerbsAndSuffixes === 'string'
+      ? JSON.parse(rawCachedAllVerbsAndSuffixes)
+      : rawCachedAllVerbsAndSuffixes;
   return cachedAllVerbsAndSuffixes;
 };
```
And this

tohuw commented:
Triage note for the maintainers: this PR appears to be **superseded** — `src/APIs/FlagsAPI.ts` and `src/APIs/RedisAPI.ts` both exist on master today (the conversion landed via #730's TS strict-mode work, later touched by #810), which is also why this branch conflicts. Unless there's something left to salvage, it can probably be closed with thanks to @shaikahmadnawaz — the goal it set out to achieve is done.

---
*This comment was created by Claude Fable 5 on behalf of [@tohuw](https://github.com/tohuw).*
